src/Base/Core/Quality/IQuality.hs

```diff
@@ -58,34 +64,38 @@ baseQuality n
 
 -- | Given an interval quality, this raises that quality by a semitone
 -- assuming that its base quality is Perfect.
-raisePerfect :: Quality -> Quality
-raisePerfect Perfect = Augmented 1
-raisePerfect (Augmented x)  = Augmented $ x + 1
-raisePerfect (Diminished 1) = Perfect
-raisePerfect (Diminished x) = Diminished $ x - 1
+raisePerfect :: Maybe Quality -> Maybe Quality
+raisePerfect (Just Perfect)        = Just $ Augmented 1
+raisePerfect (Just (Augmented x))  = iQualFrom $ Augmented $ x + 1
+raisePerfect (Just (Diminished 1)) = Just Perfect
+raisePerfect (Just (Diminished x)) = iQualFrom $ Diminished $ x - 1
+raisePerfect Nothing               = Nothing
 
 -- | Given an interval quality, this raises that quality by a semitone
 -- assuming that its base quality is Major.
-raiseMajor :: Quality -> Quality
-raiseMajor Major = Augmented 1
-raiseMajor (Augmented x)  = Augmented $ x + 1
-raiseMajor Minor          = Major
-raiseMajor (Diminished 1) = Minor
-raiseMajor (Diminished x) = Diminished $ x - 1
+raiseMajor :: Maybe Quality -> Maybe Quality
+raiseMajor (Just Major)          = Just $ Augmented 1
+raiseMajor (Just (Augmented x))  = iQualFrom $ Augmented $ x + 1
+raiseMajor (Just Minor        )  = Just Major
+raiseMajor (Just (Diminished 1)) = Just Minor
+raiseMajor (Just (Diminished x)) = iQualFrom $ Diminished $ x - 1
+raiseMajor Nothing               = Nothing
 
 -- | Given an interval quality, this lowers that quality by a semitone
 -- assuming that its base quality is Perfect.
-lowerPerfect :: Quality -> Quality
-lowerPerfect Perfect        = Diminished 1
-lowerPerfect (Diminished x) = Diminished $ x + 1
-lowerPerfect (Augmented 1)  = Perfect
-lowerPerfect (Augmented x)  = Augmented $ x-1
+lowerPerfect :: Maybe Quality -> Maybe Quality
+lowerPerfect (Just Perfect)        = Just $ Diminished 1
+lowerPerfect (Just (Diminished x)) = iQualFrom $ Diminished $ x + 1
+lowerPerfect (Just (Augmented 1))  = Just Perfect
+lowerPerfect (Just (Augmented x))  = iQualFrom $ Augmented $ x-1
+lowerPerfect Nothing               = Nothing
 
 -- | Given an interval quality, this lowers that quality by a semitone
 -- assuming that its base quality is Major.
-lowerMajor :: Quality -> Quality
-lowerMajor Major          = Minor
-lowerMajor Minor          = Diminished 1
-lowerMajor (Diminished x) = Diminished $ x + 1
-lowerMajor (Augmented 1)  = Major
-lowerMajor (Augmented x)  = Augmented $ x - 1
+lowerMajor :: Maybe Quality -> Maybe Quality
+lowerMajor (Just Major)          = Just Minor
+lowerMajor (Just Minor)          = Just $ Diminished 1
+lowerMajor (Just (Diminished x)) = iQualFrom $ Diminished $ x + 1
+lowerMajor (Just (Augmented 1))  = Just Major
+lowerMajor (Just (Augmented x))  = iQualFrom $ Augmented $ x - 1
+lowerMajor Nothing               = Nothing
```

Comment:
OK, so, this is a good example of where we might want to try to aim for functions with types of the shape `a -> m b` (that is, functions that play nicely in monadic contexts.) In this particular case, I think a more appropriate type would be `Quality -> Maybe Quality`: If things are set up correctly, we should be allowed to assume that the `Quality` we're trying to raise/lower is legitimate, but the result of raising _might not be_. Again, since the changes will be pervasive, I'll just push some edits rather than try to show it here.

Reply:
Irrelevant now that I've pushed some changed.